packages/network-controller/tests/provider-api-tests/shared-tests.ts

```diff
@@ -0,0 +1,391 @@
+            params: [dummyTransaction],
+          };
+          const expectedResult = `Unknown address - unable to sign transaction for this address: "${address}"`;
+
+          await withMockedCommunications({ providerType }, async (comms) => {
+            comms.mockNextBlockTrackerRequest({ blockNumber: '0x1' });
+
+            const promiseForResult = withNetworkClient(
+              { providerType },
+              async ({ makeRpcCall }) => makeRpcCall(request),
+            );
+
+            await expect(promiseForResult).rejects.toThrow(expectedResult);
+          });
+        });
+      });
+    });
+  });
+
+  describe('methods not included in the Ethereum JSON-RPC spec', () => {
+    describe('methods handled by middleware that returns a static result', () => {
+      const notHandledByMiddleware = [
+        { name: 'net_listening', numberOfParameters: 0, result: true },
+        {
+          name: 'web3_clientVersion',
+          numberOfParameters: 0,
+          result: 'ProviderEngine/v16.0.3/javascript',
+        },
+      ];
+      notHandledByMiddleware.forEach(({ name, numberOfParameters, result }) => {
+        describe(`method name: ${name}`, () => {
+          testsForRpcMethodWithStaticResult(name, {
+            providerType,
+            numberOfParameters,
+            result,
+          });
+        });
+      });
+    });
+
+    describe('methods not handled by middleware', () => {
+      const notHandledByMiddleware = [
+        // { name: 'eth_subscribe', numberOfParameters: 1 },
+        // { name: 'eth_unsubscribe', numberOfParameters: 1 },
+        { name: 'custom_rpc_method', numberOfParameters: 1 },
+        { name: 'net_peerCount', numberOfParameters: 0 },
+        { name: 'parity_nextNonce', numberOfParameters: 1 },
+      ];
+      notHandledByMiddleware.forEach(({ name, numberOfParameters }) => {
+        describe(`method name: ${name}`, () => {
+          testsForRpcMethodNotHandledByMiddleware(name, {
+            providerType,
+            numberOfParameters,
+          });
+        });
+      });
+    });
+
+    describe('methods that assume there is no block param', () => {
+      const assumingNoBlockParam = [
+        { name: 'eth_protocolVersion', numberOfParameters: 0 },
+      ];
+      assumingNoBlockParam.forEach(({ name, numberOfParameters }) =>
+        describe(`method name: ${name}`, () => {
+          testsForRpcMethodAssumingNoBlockParam(name, {
+            providerType,
+            numberOfParameters,
+          });
+        }),
+      );
+    });
+
+    describe('other methods', () => {
+      describe('net_version', () => {
+        it('does hit RPC endpoint to get net_version', async () => {
+          await withMockedCommunications(
+            { providerType, infuraNetwork: 'goerli', customChainId: '5' },
+            async (comms) => {
+              comms.mockRpcCall({
+                request: { method: 'net_version' },
+                response: { result: '5' },
```

Comment:
Something for a future PR: We should change this to 9999 or something, otherwise it could be a false positive if the test is wrong and the code ends up automatically matches the currently selected Infura network or chain ID.